### ISSUE TITLE
Inconsistent order of type variables in hash functions.

### ISSUE THREAD

**Issue by galenhuntington** [haskell-crypto/cryptonite#261]
Different hash functions have the two same variables ordered differently:

```haskell
hash :: (ByteArrayAccess ba, HashAlgorithm a) => ba -> Digest a
hashUpdate :: (ByteArrayAccess ba, HashAlgorithm a) => Context a -> ba -> Context a
hashWith :: (ByteArrayAccess ba, HashAlgorithm alg) => alg -> ba -> Digest alg

hashUpdates :: forall a ba . (HashAlgorithm a, ByteArrayAccess ba) => Context a -> [ba] -> Context a
digestFromByteString :: forall a ba. (HashAlgorithm a, ByteArrayAccess ba) => ba -> Maybe (Digest a)
```

There may be a deeper reason for this, but it looks unintentional.  As a result, type application use is inconsistent and possibly confusing.  E.g., `hashUpdate @ByteString @MD5` vs. `hashUpdates @MD5 @ByteString`.

Between the two choices, the second group feels more natural.  For instance, if either of these types were used--

```haskell
hash :: (HashAlgorithm a, ByteArrayAccess ba) => ba -> Digest a
hash :: forall a ba. (ByteArrayAccess ba, HashAlgorithm a) => ba -> Digest a
```

--then we could write code such as `show . hash @SHA3`, which seems more convenient and idiomatic than `show . hash @_ @SHA3`.  That said, there isn't a large difference.

What do you think?  Any change would technically be breaking, but only for those (few?) already using type applications on affected functions.

**Comment by vincenthz** [comment]
most of the APIs predates TypeApplications and Kind refinement by 4 or 5 majors version of ghc. Also `alg` should have never been a value (probably). Long time ago, no one was thinking that this unordered list of constraints would have a substantial role in type applications and would also become part of the APIs.

Ideally, I would change the APIs like what you mentioned above, but this warts will likely have to remain as is. The only path forward I can imagine is to create a shim `Crypto.Hash2` that will define things with a modern spin.

**Comment by galenhuntington** [comment]
Oops, I hadn't noticed that `alg` was a (value) argument in `hashWith`, so that one doesn't really fit with the others on my list.

Anyway, I've noticed different projects have different tolerance for backward incompatibility.  When APIs get cleaned up or refactored or modernized, some code changes are necessary.  I had to do this when, e.g., upgrading Megaparsec from 6 to 7, or for that matter upgrading to GHC 8.6.  Swapping the constraint order would be a pretty small change, but if the goal is to never break anyone's code, then indeed we're stuck with these warts.  If that's the case, I guess we can close the issue.  Or maybe Cryptonite 1.0 can be breaking?
